Add supplementary guidelines for SDK self-observability (#5135)
open-telemetry/opentelemetry-specification: specification/self-observability-supplementary-guidelines.md

the [SDK self-observability metrics semantic conventions][semconv-sdk-metrics].
The design is not inherently metrics-only - events/logs or spans describing SDK
internals may be added by future semantic conventions, so SDK implementers
should not assume the surface will remain metric-shaped.

[semconv-sdk-metrics]:
    https://opentelemetry.io/docs/specs/semconv/otel/sdk-metrics/

Once more than one signal is involved, lifecycle ordering becomes a problem. The
recording providers (`MeterProvider`, `LoggerProvider`, and potentially
`TracerProvider`) are constructed and shut down independently, so the second one
to be constructed cannot accept telemetry produced during the setup of the first
— and similarly, once a provider is shut down it can no longer accept telemetry
produced while the others are still tearing down.

For example, during startup:

* If `MeterProvider` is constructed first, self-observability *events* produced
  during its setup cannot yet flow through `LoggerProvider`, since
  `LoggerProvider` does not yet exist.
* If `LoggerProvider` is constructed first, self-observability *metrics*
  produced during its setup cannot yet flow through `MeterProvider`.

> No ordering avoids this entirely — whichever provider comes up "second" loses

Reviewer: I'm just thinking about solutions here, not suggesting updates to the PR.

Do you know if many languages implement delegating global providers?  I.e. where globals [delegate](https://github.com/open-telemetry/opentelemetry-go/blob/769a73b698d91432e57992afd922f338c1b03066/internal/global/trace.go#L45) calls to a real SDK after the SDK is Set as the global provider? I can't find any specs for it other than https://github.com/open-telemetry/opentelemetry-specification/blob/main/oteps/0005-global-init.md, but that has `status: proposed`, and I don't see any spec language for it.

Alternatively, should we specify SetMeter/Tracer/LoggerProvider methods on our providers, which would "take over" from the previous provider when set?

Author: Yes. But until the real sdk takes over, something has to hold on to the telemetry. Pausible, but definitely need some spec-work.

Reviewer: I think its totally OK if we lose observations during the brief period where one SDK has been set up (e.g. metrics), but other SDKs are being created. The key is making sure we don't lose entire instruments or meters/tracers/loggers.

Author: I don't think it is totally OK. It is entirely possible for operators/users to track the duration taken for LoggerProvider initialization - and if they report this via some `logger_provider_histogram.Record()`, and metric sdk is not initialized at that point, the measurement is gone. Metrics (like count of logs which continue to be reported throughut) which are emitted once MeterProvider is ready shouldn't be lost, except losing the count of things reported before MeterProvider is initialized.

Anyway, this guideline just list that "edge" scenarios are best-effort and nothing more.

> The key is making sure we don't lose entire instruments or meters/tracers/loggers.

+1. I haven't done a full research across implementations, but I know Go and .NET handles this well. But Rust badly handles this - if a Logger obtains a Meter early, it may get a NoOpMeter, and setting MeterProvider to a real one later won't get reflected!. It is implementation specific bug, being addressed separately).
Let me know if you think we should write something about it in this doc.

Reviewer: No, lets leave it as-is.

## Obtaining the Meter / Logger for self-observability

An SDK has two broadly different ways to acquire the `Meter` / `Logger` it
uses to emit self-observability telemetry:

* From the global provider (e.g., `GlobalMeterProvider.Get(...)`).
  Self-observability data then flows through the same pipeline as the rest of
  the user's telemetry. This is the simplest to ship and requires no additional
  configuration. The trade-off is that the user cannot easily route SDK
  self-observability separately, and the [telemetry-induced-telemetry
  concern](#avoiding-telemetry-induced-telemetry-loops) becomes more relevant
  since the SDK is emitting into its own pipeline.
* From a `MeterProvider` / `LoggerProvider` supplied explicitly by the user
  (typically via a dedicated configuration option). This makes the
  [separate pipeline](#avoiding-telemetry-induced-telemetry-loops) pattern
  viable and lets operators send SDK self-observability to a different
  backend or apply different retention/sampling. The trade-off is an
  additional configuration surface, and a fallback decision when no
  provider is supplied (e.g., fall back to global, or emit nothing).

> Both choices are valid and depend on the SDK's audience and how strongly it

Reviewer: This isn't an either-or, right?  At least in Go the best practice for instrumentation libraries is to accept a user-supplied Provider, and fall-back to the global if the user doesn't provide one.

Author: agree. tweaked the wording to reflect this.

wants to enable separate routing.

## Avoiding telemetry-induced-telemetry loops

When the SDK emits self-observability data through its own telemetry pipeline,
there is a risk of infinite recursion — an exporter that records a metric on
every export triggers another export, which records another metric, and so on.

Patterns SDKs can use to prevent such loops:

> * Use a dedicated `MeterProvider` (or `LoggerProvider`) for self-observability

Reviewer: I'm not sure this is an issue for metrics...  IIUC, the issue for tracers and loggers is that the volume of traces/logs they produce themselves is based on the volume of traces/logs they handle.  For metrics, you aggregate observations, so 10x more measure calls doesn't translate into 10x the export volume in the same way.

Author: That is correct. (https://github.com/open-telemetry/opentelemetry-rust/issues/2882 also discussed this).
Rewording to reflect this as a log/trace concern.

    https://opentelemetry.io/docs/specs/semconv/otel/sdk-metrics/

Once more than one signal is involved, lifecycle ordering becomes a problem. The
recording providers (`MeterProvider`, `LoggerProvider`, and potentially
`TracerProvider`) are constructed and shut down independently, so the second one
to be constructed cannot accept telemetry produced during the setup of the first
— and similarly, once a provider is shut down it can no longer accept telemetry
produced while the others are still tearing down.

For example, during startup:

* If `MeterProvider` is constructed first, self-observability *events* produced
  during its setup cannot yet flow through `LoggerProvider`, since
  `LoggerProvider` does not yet exist.
* If `LoggerProvider` is constructed first, self-observability *metrics*
  produced during its setup cannot yet flow through `MeterProvider`.

No ordering avoids this entirely — whichever provider comes up "second" loses
the window before it exists, and whichever is shut down "first" loses the window
after it is gone. Self-observability telemetry at the edges of the SDK lifecycle
is therefore inherently best-effort; the strategy for handling it is left to the

> SDK.

Reviewer 2: So in otel java, here's how we solve this:

- Our programmatic config APIs for configuring internal telemetry accept a `Supplier<MeterProvider>` rather than a concrete instance. For example: https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorBuilder.java#L161
- All internal telemetry code paths initialize lazily, allowing the `Supplier<MeterProvider>` to swap implementations once the resolved MeterProvider becomes available without locking in a noop MeterProvider, and a noop Meter and noop instruments. For example, [in this instrumentation](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/SemConvSpanProcessorInstrumentation.java#L57), we need to increment a counter. Rather than doing the typical thing and resolving the meter / instrument at initialization time, we call `processedSpan()` which initializes the meter / counter just in time for the first measurement recorded.

Author: Thanks! I think it lines up with what I mentioned in the thread above with @dashpole (https://github.com/open-telemetry/opentelemetry-specification/pull/5135/changes#r3373670821): lazy Supplier<MeterProvider> (and equivalent in other SDKs) is exactly what avoids permanently binding to a noop Meter, so ongoing measurements after startup flow correctly. 

The "best-effort at edges" wording in the doc is specifically about the values produced inside the gap (e.g., recording the duration of LoggerProvider init OR OTLPLogExporter initialization duration etc before MeterProvider exists) — those are lost regardless. So the two are complementary rather than in conflict. Let me know we should tweak any wording here.

Author: @jack-berg Gentle ping on this.

For self-observability logs/events specifically, if the SDK already emits
diagnostics through a non-OpenTelemetry path — the language's native logging
facility, a commonly-used ecosystem logging library (e.g., `tokio-tracing` in
Rust), or in the simplest case direct writes to stdout/stderr — that path is a
natural fit for events emitted before `LoggerProvider` is installed or after it
has been shut down. It is typically available throughout the process lifetime
and has few external dependencies that can fail.

> ## Obtaining the Meter / Logger for self-observability

Reviewer: I wonder if this would make more sense in instrumentation libraries guidelines?

I think it would also not be a bad thing to link from this document to the general guidelines. Self-observability instrumentation is just instrumentation.

Author: Good point. But I don't think we have such a guideline doc for instrumentation libraries. (let me know if I missed finding it!).

I can open an issue to track this either in spec repo or semantic conventions (I think spec is better home for that?)

Reviewer: Ah, I was thinking of https://opentelemetry.io/docs/concepts/instrumentation/libraries/, but that isn't actually a spec, so i'm not sure if we can link it.

Reviewer: Given this is just supplementary guidelines, I actually think linking to user documentation for instrumentation best practices is fine.  I would love to see this topic covered there eventually, but i'm OK just merging this as-is.

Author: https://github.com/open-telemetry/opentelemetry-specification/issues/5148 opened this. If there is agreement on it, then it should be relatively easy to make happen by combining existing contents. And then I can link from this doc.

span the SDK produces while handling a log or span can itself cause more logs
or spans to be produced, leading to unbounded recursion. Metrics are less
affected in practice.

Patterns SDKs can use to prevent such loops:

* Use a dedicated `LoggerProvider` (or `TracerProvider`) for self-observability
  that is isolated from the user's pipeline, so self-observability telemetry
  does not feed back into it.
* Use the OpenTelemetry `Context` to carry a flag marking code as running inside
  the SDK's own pipeline, and skip self-observability recording when the flag is
  set. There is no standardized spec for this today (tracked in
  [open-telemetry/opentelemetry-specification#530](https://github.com/open-telemetry/opentelemetry-specification/issues/530));
  in the meantime, several SDKs implement it independently:
  * .NET:
    [`SuppressInstrumentationScope`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/core-1.15.3/src/OpenTelemetry/SuppressInstrumentationScope.cs#L12)
  * Rust:
    [`Context::enter_telemetry_suppressed_scope`](https://github.com/open-telemetry/opentelemetry-rust/blob/opentelemetry-0.32.0/opentelemetry/src/context.rs#L410)
  * Python:

>     [`_SUPPRESS_INSTRUMENTATION_KEY`](https://github.com/open-telemetry/opentelemetry-python/blob/v1.42.1/opentelemetry-api/src/opentelemetry/context/__init__.py#L152)

Reviewer 2: Here's the java equivalent: https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/impl/InstrumentationUtil.java#L23-L24

Author: thanks. added to the list. (Pretty sure now someone will ask - if all sdks need it, why is it not int he spec already 🤣 )

## Treat self-observability like any other SDK feature for stability

SDK self-observability is an SDK feature and is subject to the SDK's normal
[stability guarantees](versioning-and-stability.md) — no weaker, no different.

This has three consequences:

1. Any in-development or experimental metric, attribute, or semantic must be

>    opt-in.

Reviewer 2: ☝️

Author: Stable-by-default :)

# Self-Observability Supplementary Guidelines

Note: this document is NOT a spec, it is provided to support the
[Self-Observability](self-observability.md) specification, it does NOT add any
extra requirements to the existing specifications.

<details>
<summary>Table of Contents</summary>

<!-- toc -->

- [Scope of signals and lifecycle ordering](#scope-of-signals-and-lifecycle-ordering)
- [Obtaining the Meter / Logger for self-observability](#obtaining-the-meter--logger-for-self-observability)
- [Avoiding telemetry-induced-telemetry loops](#avoiding-telemetry-induced-telemetry-loops)
- [Treat self-observability like any other SDK feature for stability](#treat-self-observability-like-any-other-sdk-feature-for-stability)

<!-- tocstop -->

</details>

## Scope of signals and lifecycle ordering

SDK self-observability is currently expressed primarily as metrics, defined in
the [SDK self-observability metrics semantic conventions][semconv-sdk-metrics].

> The design is not inherently metrics-only - events/logs or spans describing SDK

Reviewer 3: ```suggestion
The design is not inherently metrics-only. Events or spans describing SDK
```

internals may be added by future semantic conventions, so SDK implementers
should not assume the surface will remain metric-shaped.

[semconv-sdk-metrics]: https://opentelemetry.io/docs/specs/semconv/otel/sdk-metrics/

Once more than one signal is involved, lifecycle ordering becomes a problem. The
recording providers (`MeterProvider`, `LoggerProvider`, and potentially
`TracerProvider`) are constructed and shut down independently, so the second one

> to be constructed cannot accept telemetry produced during the setup of the first

Reviewer 3: ```suggestion
to be constructed cannot accept telemetry produced during the setup of the first.
```

> — and similarly, once a provider is shut down it can no longer accept telemetry

Reviewer 3: ```suggestion
Similarly, once a provider is shut down it can no longer accept telemetry
```

produced while the others are still tearing down.

For example, during startup:

* If `MeterProvider` is constructed first, self-observability *events* produced
  during its setup cannot yet flow through `LoggerProvider`, since
  `LoggerProvider` does not yet exist.
* If `LoggerProvider` is constructed first, self-observability *metrics*
  produced during its setup cannot yet flow through `MeterProvider`.

No ordering avoids this entirely — whichever provider comes up "second" loses
the window before it exists, and whichever is shut down "first" loses the window
after it is gone. Self-observability telemetry at the edges of the SDK lifecycle
is therefore inherently best-effort; the strategy for handling it is left to the
SDK.

> For self-observability logs/events specifically, if the SDK already emits

Reviewer 3: ```suggestion
For self-observability events specifically, if the SDK already emits
```

  (typically via a dedicated configuration option). This makes the
  [separate pipeline](#avoiding-telemetry-induced-telemetry-loops) pattern
  viable and lets operators send SDK self-observability to a different
  backend or apply different retention/sampling. The trade-off is an
  additional configuration surface, and a fallback decision when no
  provider is supplied (e.g., fall back to global, or emit nothing).

Both choices are valid and depend on the SDK's audience and how strongly it
wants to enable separate routing.

These two approaches can also be combined: an SDK can accept an explicit
`MeterProvider` / `LoggerProvider` and fall back to the global when none is
supplied. This is a common pattern in instrumentation libraries and gives
operators the option to route SDK self-observability separately without
forcing additional configuration on users who do not need it.

## Avoiding telemetry-induced-telemetry loops

When the SDK emits self-observability data through its own telemetry pipeline,
the data it emits can in turn be processed by that same pipeline, creating a
feedback loop. This is primarily a concern for events and traces: each event or

> span the SDK produces while handling a event or span can itself cause more events

Reviewer 3: ```suggestion
span the SDK produces while handling an event or span can itself cause more events
```

> the data it emits can in turn be processed by that same pipeline, creating a
> feedback loop. This is primarily a concern for logs and traces: each log or
> span the SDK produces while handling a log or span can itself cause more logs
> or spans to be produced, leading to unbounded recursion. Metrics are less

Reviewer 3: ```suggestion
the data it emits can in turn be processed by that same pipeline, creating a
feedback loop. This is primarily a concern for events and traces: each event or
span the SDK produces while handling a event or span can itself cause more events
or spans to be produced, leading to unbounded recursion. Metrics are less
```
